### PR TITLE
Subscriptions authentication

### PR DESCRIPTION
**Pull request by angrykoala**
# Description

Adds authentication checks to subscriptions, throwing an error to connect if not authenticated.

Support for `isAuthenticated` and `allowUnauthenticated`

This PR depends on https://github.com/neo4j/graphql/pull/1350

### CODE REVIEW

File: packages/graphql/src/types.ts [hunk 1]
```diff
@@ -70,10 +70,10 @@ export interface AuthRule extends BaseAuthRule {
     operations?: AuthOperations[];
 }
 
-export type Auth = {
+export interface Auth {
```
Comment (angrykoala) [inline packages/graphql/src/types.ts:73]: Changed to interface to support implementing it in classes

File: packages/graphql/src/utils/get-token.ts [hunk 2]
```diff
@@ -63,3 +62,11 @@ export function getToken(context: Context): string | undefined {
 
     return token;
 }
+
+export function parseBearerToken(bearerAuth: string): string {
+    const token = bearerAuth.split("Bearer ")[1];
+    if (!token) {
+        debug("Authorization header was not in expected format 'Bearer <token>'");
```
Comment (darrellwarde) [inline packages/graphql/src/utils/get-token.ts:69]: **question:** surely this is just going to break down the line if we don't throw and error here?

File: packages/graphql/tests/e2e/subscriptions/auth/authentication.int.test.ts [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+        let server: TestGraphQLServer;
+        let wsClient: WebSocketTestClient;
+
+        beforeAll(async () => {
+            const typeDefs = `
+            type ${typeMovie} {
+                title: String!
+            }
+
+            extend type ${typeMovie} @auth(rules: [{ isAuthenticated: true, allowUnauthenticated: true, operations: [SUBSCRIBE] }])
```
Comment (darrellwarde) [inline packages/graphql/tests/e2e/subscriptions/auth/authentication.int.test.ts:484]: Perhaps out of the scope of this PR, but I feel like combining `isAuthenticated` and `allowUnauthenticated` shouldn't be allowed, and `allowUnauthenticated` should only be allowed to be combined with _authorization_ rules. We should discuss this!

**Review by darrellwarde — DISMISSED** [review]
Sorry for the slow review, this looks good! I think the error message perhaps needs to be a bit clearer, but other than that, just ESLint violations to address.

File: packages/graphql/src/utils/get-token.ts (hunk 2 above)
Comment (angrykoala) [reply packages/graphql/src/utils/get-token.ts:69]: Possibly, this logic is actually the same as our current token logic for normal requests (same file, line 58).

I'm not sure what the fallback is for or its implications, but I think that falls under authentication improvements task we have pending to do

File: packages/graphql/tests/e2e/subscriptions/auth/authentication.int.test.ts (hunk 3 above)
Comment (angrykoala) [reply packages/graphql/tests/e2e/subscriptions/auth/authentication.int.test.ts:484]: We need to discuss, isn't the point of allow unauthenticated to support whitelisting?

**Pull request merged**